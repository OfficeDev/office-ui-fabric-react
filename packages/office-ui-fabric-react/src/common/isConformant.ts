// eslint-disable-next-line import/no-extraneous-dependencies -- this file is for testing
import { isConformant as isConformantBase, IsConformantOptions } from '@fluentui/react-conformance';

export function isConformant(testInfo: IsConformantOptions) {
  const defaultOptions = {
    disabledTests: ['has-docblock'],
  };

  isConformantBase(defaultOptions, testInfo);
}
