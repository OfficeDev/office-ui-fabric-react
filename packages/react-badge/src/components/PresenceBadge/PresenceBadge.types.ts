import { BadgeProps, BadgeState } from '../Badge/index';

/**
 * {@docCategory PresenceBadge}
 */
export type PresenceBadgeStatus = 'busy' | 'oof' | 'away' | 'available' | 'offline';

/**
 * {@docCategory PresenceBadge}
 */
export interface PresenceBadgeProps extends Omit<BadgeProps, 'shape' | 'appearance'> {
  /**
   * A PresenceBadge can represent several status
   * @defaultvalue available
   */
  status?: PresenceBadgeStatus;
  /**
   * A PresenceBadge can represent status of someone out of the office
   * @defaultvalue true
   */
  inOffice?: boolean;
}

/**
 * {@docCategory Badge}
 */
export interface PresenceBadgeState extends BadgeState {
  /**
   * A PresenceBadge can represent several status
   * @defaultvalue available
   */
  status: PresenceBadgeStatus;
  /**
   * A PresenceBadge can represent status of someone out of the office
   * @defaultvalue true
   */
  inOffice: boolean;
}
