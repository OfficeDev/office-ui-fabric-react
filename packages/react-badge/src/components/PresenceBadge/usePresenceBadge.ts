import * as React from 'react';
import { makeMergeProps } from '@fluentui/react-utilities';
import { PresenceBadgeProps, PresenceBadgeState } from './PresenceBadge.types';
import { useBadge, BadgeProps } from '../Badge/index';

/**
 * Consts listing which props are shorthand props.
 */
export const presenceBadgeShorthandProps: (keyof PresenceBadgeProps)[] = ['icon'];

const mergeProps = makeMergeProps<PresenceBadgeState>({ deepMerge: presenceBadgeShorthandProps });

/**
 * Returns the props and state required to render the component
 */
export const usePresenceBadge = (
  props: PresenceBadgeProps,
  ref: React.Ref<HTMLElement>,
  defaultProps?: PresenceBadgeProps,
): PresenceBadgeState => {
  const state = useBadge(
    props,
    ref,
    mergeProps(
      {
        size: 'small',
        status: 'available',
        outOfOffice: false,
      },
      defaultProps,
    ) as BadgeProps,
  ) as PresenceBadgeState;

  return state;
};
