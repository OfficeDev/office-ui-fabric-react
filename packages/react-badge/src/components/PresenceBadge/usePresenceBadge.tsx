import * as React from 'react';
import { makeMergeProps } from '@fluentui/react-utilities';
import { PresenceBadgeProps, PresenceBadgeState, PresenceBadgeStatus } from './PresenceBadge.types';
import { useBadge, BadgeProps } from '../Badge/index';
import {
  SkypeMinusIcon,
  SkypeClockIcon,
  SkypeArrowIcon,
  SkypeCheckIcon,
  CancelIcon,
} from './DefaultPresenceBadgeIcons';

/**
 * Consts listing which props are shorthand props.
 */
export const presenceBadgeShorthandProps: (keyof PresenceBadgeProps)[] = ['icon'];

const mergeProps = makeMergeProps<PresenceBadgeState>({ deepMerge: presenceBadgeShorthandProps });

const iconMap: Record<PresenceBadgeStatus, JSX.Element | null> = {
  busy: <SkypeMinusIcon />,
  available: <SkypeCheckIcon />,
  away: <SkypeClockIcon />,
  offline: <CancelIcon />,
  outOfOffice: <SkypeArrowIcon />,
  doNotDisturb: null,
};

/**
 * Returns the props and state required to render the component
 */
export const usePresenceBadge = (
  props: PresenceBadgeProps,
  ref: React.Ref<HTMLElement>,
  defaultProps?: PresenceBadgeProps,
): PresenceBadgeState => {
  const state = useBadge(
    props,
    ref,
    mergeProps(
      {
        size: 'small',
        status: 'available',
        outOfOffice: false,
        icon: { as: 'span' },
      },
      defaultProps,
    ) as BadgeProps,
  ) as PresenceBadgeState;

  if (!state.icon?.children) {
    state.icon!.children = iconMap[state.status];
  }

  return state;
};
