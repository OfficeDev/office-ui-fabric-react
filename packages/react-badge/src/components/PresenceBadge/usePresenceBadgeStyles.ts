import { ax, makeStyles } from '@fluentui/react-make-styles';
import { PresenceBadgeState } from './PresenceBadge.types';
import { useBadgeStyles } from '../../Badge';

export const usePresenceBadgeRootStyles = makeStyles<PresenceBadgeState>([
  [
    null,
    {
      padding: 0,
    },
  ],
  [
    s => s.status === 'busy' || s.status === 'doNotDisturb',
    theme => ({
      backgroundColor: theme.alias.color.cranberry.background3,
      borderColor: theme.alias.color.cranberry.background3,
    }),
  ],
  [
    s => s.status === 'away',
    theme => ({
      backgroundColor: theme.global.palette.yellow.primary,
      borderColor: theme.global.palette.yellow.primary,
    }),
  ],
  [
    s => s.status === 'available',
    theme => ({
      backgroundColor: theme.global.palette.lightGreen.primary,
      borderColor: theme.global.palette.lightGreen.primary,
    }),
  ],
  [
    s => s.status === 'offline',
    theme => ({
      backgroundColor: 'transparent',
      color: theme.global.palette.grey[38],
      borderColor: theme.global.palette.grey[38],
    }),
  ],
  [
    s => s.status === 'outOfOffice',
    theme => ({
      backgroundColor: theme.alias.color.neutral.neutralBackground1,
      color: theme.global.palette.magenta.primary,
      borderColor: theme.global.palette.magenta.primary,
    }),
  ],
  [
    s => s.outOfOffice,
    {
      backgroundColor: 'transparent',
    },
  ],
  [
    s => s.outOfOffice && s.status === 'available',
    theme => ({
      color: theme.global.palette.lightGreen.primary,
    }),
  ],
  [
    s => s.outOfOffice && (s.status === 'busy' || s.status === 'doNotDisturb'),
    theme => ({
      color: theme.global.palette.red.primary,
    }),
  ],
  [
    s => s.outOfOffice && s.status === 'away',
    theme => ({
      color: theme.global.palette.yellow.primary,
    }),
  ],
]);

/**
 * Applies style classnames to slots
 */
export const usePresenceBadgeStyles = (state: PresenceBadgeState) => {
  state.className = ax(usePresenceBadgeRootStyles(state), state.className);

  return useBadgeStyles(state);
};
