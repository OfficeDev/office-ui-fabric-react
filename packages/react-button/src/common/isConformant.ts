import { isConformant as baseIsConformant, IsConformantOptions } from '@fluentui/react-conformance';

export function isConformant(testInfo: Omit<IsConformantOptions, 'componentPath'>) {
  const defaultOptions = {
    disabledTests: ['has-docblock', 'kebab-aria-attributes', 'as-*'],
    componentPath: module!.parent!.filename.replace('.test', ''),
  };

  baseIsConformant(defaultOptions, testInfo);
}
