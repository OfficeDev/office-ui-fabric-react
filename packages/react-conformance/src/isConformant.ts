import { TestObject, TestingOptions } from './types';
import { withCustomConfig } from 'react-docgen-typescript';
import * as path from 'path';
import { defaultTests } from './defaultTests';

export function isConformant(testInfo: TestingOptions, extraTests?: TestObject) {
  const { componentPath, displayName } = testInfo;
  const tsconfigPath = path.join(process.cwd(), 'tsconfig.json');
  const parser = withCustomConfig(tsconfigPath, {});
  const components = parser.parse(componentPath);
  const mainComponents = components.filter(comp => comp.displayName === displayName);

  if (mainComponents.length === 1) {
    const componentInfo = mainComponents[0];

    for (const test of Object.keys(defaultTests)) {
      defaultTests[test](componentInfo, testInfo);
    }
    if (extraTests) {
      for (const test of Object.keys(extraTests)) {
        extraTests[test](componentInfo, testInfo);
      }
    }
  } else {
    if (components.length === 0) {
      throw Error('No exported components in path: ' + componentPath);
    } else {
      throw Error(
        `No component with name '${displayName}' was found, these are the exported components: {${components
          .map(component => {
            return component.displayName;
          })
          .toString()}} in path ${componentPath}`,
      );
    }
  }
}
