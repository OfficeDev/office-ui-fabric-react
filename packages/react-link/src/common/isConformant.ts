import { isConformant as baseIsConformant, IsConformantOptions } from '@fluentui/react-conformance';

export function isConformant(testInfo: Omit<IsConformantOptions, 'componentPath'>) {
  const defaultOptions = {
    disabledTests: [
      `has-docblock`,
      `kebab-aria-attributes`,
      `as-renders-fc`,
      `as-renders-react-class`,
      `as-passes-as-value`,
    ],

    componentPath: module!.parent!.filename.replace('.test', ''),
  };

  baseIsConformant(defaultOptions, testInfo);
}
