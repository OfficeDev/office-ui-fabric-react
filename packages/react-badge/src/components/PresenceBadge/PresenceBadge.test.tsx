import * as React from 'react';
import { PresenceBadge } from './PresenceBadge';
import * as renderer from 'react-test-renderer';
import { ReactWrapper } from 'enzyme';
import { isConformant } from '../../common/isConformant';

describe('Presence', () => {
  isConformant({
    Component: PresenceBadge,
    displayName: 'Presence',
  });

  let wrapper: ReactWrapper | undefined;

  afterEach(() => {
    if (wrapper) {
      wrapper.unmount();
      wrapper = undefined;
    }
  });

  /**
   * Note: see more visual regression tests for Badge in /apps/vr-tests.
   */
  it('renders a default state', () => {
    const component = renderer.create(<PresenceBadge />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
