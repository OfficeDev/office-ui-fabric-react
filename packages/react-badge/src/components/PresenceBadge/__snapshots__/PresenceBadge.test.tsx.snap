// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PresenceBadge renders a default state 1`] = `
<div
  className=" "
/>
`;
